Wayland: Implement game embedding

deralmas opened the pull request:
**As per tradition, I summarized this description and updated it with all the new developments ever since I opened this PR. If you want to see the original, check out the edit history.**

This PR implements game embedding on Wayland, using a novel experimental approach.

The code is basically feature complete, with the exception of tablet support, due to various quirks.

![image](https://github.com/user-attachments/assets/114693e4-51e4-48f4-bf6b-737284ea73dc)

## Testing

Obviously, the editor must be running with the `Prefer Wayland` option on in the editor settings. 

(Previous iterations of this PR required the project to specify the Wayland driver too, this is not the case anymore) 

When debugging, you can uncomment `#define WAYLAND_EMBED_DEBUG_LOGS_ENABLED`, which will completely _blast_ your terminal with a lot of information, ranging from hex dumps to various debug notes. Because of this, **whenever it is defined** I recommend `tee`ing the whole thing to a file. I usually run it like this:

```shell
path/to/godot --verbose 2>&1 | tee proxy.log
```

These logs are extremely verbose on purpose and, if you need to trim them down when sharing, please keep a full copy as I might need more data later.

That said, please be aware that: as with any protocol dump, **these logs contain every event, including key presses sent to the editor/game** (not the whole OS), just like `WAYLAND_DEBUG=1`. I'm open to receiving logs privately if that makes you feel safer.

## How it works

This code acts as a shim  between the user's compositor and the editor (templates and the project managers skip it for stability). It intercepts all Wayland messages and manipulates them to achieve things not normally possible such as redirecting a window to a subsurface, all without modifying the embedded client. This shim is referred as the "embedder" in the code.

The embedding client can access a "fake" Wayland global offered by the shim with various methods to handle all embedded clients.

In practice it "merges" multiple clients into one single connection, which has various pro and cons. I'll explain below why I chose this approach.

Additionally I tried to make the code as "automatic" as possible; In most instances all it takes is to make the embedder aware of an interface and it takes care of any requests through a generic handler.

I'll document the approach properly at least after this PR gets ready but here's a very quick and rambly summary of what comes to mind:

<details>

This is basically a compositor without all the actual logic (except when we need to emulate stuff for stuff like fake windows obviously). The first client to connect is designated as the "main" client, which creates actual windows, while all new clients get embedded as soon as they make a new toplevel by intercepting the request. Everything else still gets created (buffers, etc.), which means that we can simply take the buffer that was meant for a toplevel and place it into a subsurface. We need to emulate just enough of a toplevel to keep the client happy and both the compositor and the client will do all the heavy lifting for us.

The most basic building block in this project is that there are various "ID spaces": since we're handling multiple clients, and the Wayland protocol enforces reusing IDs (with an assertion in libwayland), we need a "global" ID space (with all the "real" objects from the actual compositor) and a per-client "local" ID space, with relative mappings between both.

I made sure to add plenty of utility methods and some wrappers to hopefully make handling them both easier.

We also need to manage the real compositor's global objects (what I call "registry objects" to disambiguate) since we offer the control object (we already have an IPC after all, why reinvent the wheel? :P ).

The most annoying part was a certain mechanism I had to implement for compatibility with older compositors because it looks like you couldn't cleanup certain objects (mainly core registry globals) until relatively recently.

Thus a compatibility mechanism was born: "registry globals instances", basically if we detect that an object does not have a destructor method we instance it once and reuse it by creating fake objects. It works surprisingly well although it requires some extra state tracking for some specific objects.

It's not implemented perfectly (there are some missing edge cases) but as all globals now implement destructor methods hopefully it should kick in less and less over time.

It definitely adds complexity and in all honestly I should've considered just disabling the feature on old compositors but since I was still experimenting I did not understand fully the issue yet :sweat_smile:

</details>

## How it's written

It started from a standalone C prototype which got grafted into the Godot codebase. I refactored it extensively since then to follow Godot's style and data structures. If something is weird, that's probably why.

This code went through _lots_ of iterations due to its experimental nature. I'm not super proud about how the data structures are laid out and the names are kinda eh. I'll keep improving it as much as possible but I don't expect it to become perfect by any means. There are a lot of moving parts and stuff works, so I'm not going to break it if I'm not forced to :P

I plan to eventually create a third-party library that implements this approach or at the very least to fully document the current architecture. Hopefully that will give further insights into simplifying it further.

There are also some naively written parts and dumb things in general but luckily performance seems to not be an issue: https://github.com/godotengine/godot/pull/107435#issuecomment-3186283143

## Why this approach

The "canonical" approach for Wayland embedding would be to implement a more or less full fledged compositor[^canonical-approach], usually with a library like wlroots.

I initially looked into it but it looks like all compositor libraries (understandably) expect that you will be running them under less capable environments (DRM/libinput, X11, older compositors...). Thus we would need to normally re-implement basically every single event handler as a pass-through, so I started considering handling the issue differently. 

The advantages overall are very interesting:

 - Maintainability: the idea is that most Wayland protocols can simply "pass through" so, by introspecting into the generated headers, all you need to do to "implement" a protocol is adding its static interface pointer into a list.

 - Size and simplicity: the actual embedder is substantially smaller than your standard compositor, comments and all. By doing our trick, we get _a lot_ of stuff "for free".

 - Integration: at least wlroots renders everything on a single buffer (I assume all compositor libraries do). We instead simply "trick" the secondary clients into rendering onto a subsurface. This means that we don't need a renderer and most importantly that popups and other extra surfaces can go outside of the embedded window, which I think is a _huge_ plus.

 - Rendering performance?: I'm not sure how to verify this but I suppose that because we're not doing an extra step(?) of rendering and simply redirecting the embedded client to different surfaces that there should be minimal overhead.

There are also disadvantages I should point out:

 - Everything runs under the embedder, editor included. This means that, if the embedded game does "naughty things" and the real compositor disconnects it, it will actually kick both the editor and the game. This is not critical as we trust the client and could do the kicking ourselves but that's note-worthy. Also note that the embedder is enabled only with the editor, the project manager and non-embedded projects will run "on bare metal" right away.

 - This only works for Wayland clients. Honestly, this is not that big of a deal because there are drop-in solutions like [xwayland-satellite](https://github.com/Supreeeme/xwayland-satellite) that can handle xwayland for us if we ever feel the need. Some compositors like niri went with the same approach: https://github.com/YaLTeR/niri/wiki/Xwayland **Update**: This is even less of an issue now as we simply pass `--display-driver wayland` to the embedded game.

 - You need to register every protocol used, even low level ones. We thus need to add and track extra XML definitions for stuff used by the WSI. It's one line in the actual code but still, worthy of note.

 - *You need to register every protocol used.* Yes, this makes the barrier of entry to contribution ever so slightly high but if the protocol does not touch the few things the embedder "overrides" it should be just a line in the embedder code.

## Yea.

<details>

<summary> Old conclusion (It felt wrong to change it) </summary>

> This was meant to be short to not waste too much of my energy but that's apparently what short looks like lmao. It did not help that I accidentally lost half of this description right before sending it and had to rewrite it from scratch.
> 
> Really, there's _a lot_ I've omitted, such as the constraints the wire format imposes and the actual architecture of this code. I'll document everything (and more) once it's in a stable state. This is only the beginning.
> 
> You have no idea how much work went into getting at this point. It was _brutal_. I won't lie, I might've underestimated it and that might have contributed to my... lack of energy. I don't think it was the only reason though.
> 
> This is the most experimental thing I've ever done: I'm still waiting for the huge wall into which everything will crash spectacularly but right now I've went through every single hurdle, which makes it quite hard to keep the enthusiasm in control. If I actually nailed it, it should be all downhill from now on.
> 
> I really believe this might be _it_, otherwise I would not have spent 3+ hours writing this wall of text almost twice. I want to get this out, to let everybody know what has been brewing behind the scenes and give them the chance to try it instead of sitting on this code waiting for the worst to happen.
> 
> However it goes, I learnt a lot and had a blast playing with the wire itself.
> 
> Thank you for your patience.
> 

</details>

The original conclusion is outdated in some ways but I'm keeping it because it sums exactly the feelings I had while working on this thing. I don't want to get personal so I'll limit myself by saying that those annoying times for unrelated reasons and I went pretty darn emotional for a PR description. I can't write normal PRs for the life of me, sorry xD

That said, as per the original text, this was way more complex than I thought but I still genuinely believe that this makes more sense than using a compositor library like wlroots. In both cases you're writing a compositor and that's a hard thing to do, there's no way around that.

Hopefully I'm right :P

[^canonical-approach]: To be completely fair, another option mentioned in the Wayland docs themselves would be to pass the game buffer out-of-band like with the MacOS embedder. I personally think that in our particular case, given Wayland's architecture, the approach I'm implementing is the right choice.

cg9999 commented:
Thank you, seems to be working great!

I'm using Cosmic Desktop on Arch. A few warning/error messages seen so far:

`WARNING: FIFO protocol not found! Frame pacing will be degraded.
     at: init (platform/linuxbsd/wayland/wayland_thread.cpp:4430)

ERROR: Condition "p_window_id != MAIN_WINDOW_ID" is true. Returning: INVALID_SCREEN
   at: window_get_current_screen (platform/linuxbsd/wayland/display_server_wayland.cpp:1053)

When stopping game:
ERROR: Can't read message header: Connection reset by peer
   at: handle_sock (platform/linuxbsd/wayland/snooper.cpp:1951)

ERROR: Condition "((size_t)head_rec) != vec.iov_len" is true. Returning: false
   at: handle_sock (platform/linuxbsd/wayland/snooper.cpp:1952)`

I can upload full logs if any of these is interesting.

ArchercatNEO commented:
Interestingly I can't seem to embed at all on KDE Plasma, NixOS.

![Screenshot_20250614_092610](https://github.com/user-attachments/assets/761b2d5d-5eaf-4e2d-bdf6-388128459039)

There aren't any relevant errors opening the window and closing the window only says the client disconnected.
Something even more strange is that when closing the editor I get spammed with `ERROR: Condition "((size_t)head_rec) != vec.iov_len" is true. Returning: false` and then we crash with an index out of bounds. We were already closing the editor so it crashing isn't as big as it crashing while stuff is running but it is a significant thing

ArchercatNEO commented:
@Riteo unfortunately it seems there are more issues and I still can't embed

There was nothing wrong with the window it just wasn't embedding (I hadn't set up the images to actually fit in the screen)
I am on NixOS but have my repo set to `nixos-unstable` which should mean I'm on a rolling release too (my Plasma is 6.3.5)
I have both x11 and libdecor disabled in the compilation flags which may mean something

Here's what I mean by the window not embedding, there's nothing wrong with the content but it just isn't in the game tab
[Screencast_20250615_102539.webm](https://github.com/user-attachments/assets/f24e1cfe-d6bc-46de-af52-b396a4521012)

Also a *MUCH* weirder bug, despite not having libdecor both windows have decorations but this warning is still raised somehow. This didn't happen last time I tried it but it does now.
![image](https://github.com/user-attachments/assets/69ed73ea-40c0-491d-a575-626544f85c3b)

ArchercatNEO commented:
No clue what changed but it works now.
I ran the editor under gdb and it embedded right so I tried again without gdb and got embedding.

![Screenshot_20250615_105613](https://github.com/user-attachments/assets/1a0d507e-c0d7-487c-9999-3080d6d97587)

ArchercatNEO commented:
Random crash with this log. The embeded window wasn't even open so this is something with the main window itself
```
[PROXY]  === START PACKET ===
[PROXY] Received bytes: 710000000100100000bd010000750000
[PROXY] dir: compositor, id: 0x71, bytes: 16, opcode: 1
[PROXY] Client: 0.
ERROR: No object found for r0x71
   at: handle_msg_info (platform/linuxbsd/wayland/snooper.cpp:1911)
ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
   at: propagate_notification (scene/main/node.cpp:2543)
```

ArchercatNEO commented:
Looking through the protocols used I noticed you added the `linux-explicit-synchronization` protocol but according to wayland.app it's been deprecated in favour of  `linux-drm-syncobj`. Would this be like `wayland-drm` where the WSI in older vulkan versions uses the unstable protocol so we have to expose it? If so shouldn't we also add `linux-drm-syncobj` for newer versions of WSI?

ArchercatNEO commented:
Sorry to do this in 2 parts but I looked to mesa's master branch to see which protocols they use in https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/loader/meson.build#L8-16

(Transcripted)
```
'fifo-v1': 'staging/fifo/fifo-v1.xml',
'commit-timing-v1': 'staging/commit-timing/commit-timing-v1.xml',
'linux-dmabuf-unstable-v1': 'unstable/linux-dmabuf/linux-dmabuf-unstable-v1.xml',
'presentation-time': 'stable/presentation-time/presentation-time.xml',
'tearing-control-v1': 'staging/tearing-control/tearing-control-v1.xml',
'linux-drm-syncobj-v1': 'staging/linux-drm-syncobj/linux-drm-syncobj-v1.xml',
'color-management-v1': 'staging/color-management/color-management-v1.xml',
```

I am guessing that these are the protocols we need to relay down to the embedded process since wsi/egl will attempt to use protocols even if we don't. It also seems that regardless of egl/vulkan these are all the protocols we need to worry about.

Of these it seems we are missing `commit-timing-v1`, `presentation-time`, `tearing-control-v1` and `color-management-v1`.
We will actually make use of `color-management-v1` but in a different PR, so whether it gets added here or in that one mostly comes down to which PR gets merged first.

AndreaMonzini commented:
Thank you  @Riteo for the work !

Questions:
- is this PR present in the last public Godot 4.5 beta 1 ?  ( for an easy sharing test )
- have you considered a different approach like the https://github.com/godotengine/godot/pull/105884 ?
  For what i understand in this case there is not a window management hack.
 
In any case it seems really a lot of work and i will test as soon as possible !

ArchercatNEO commented:
Yes, from the footnote in the first post
> To be completely fair, another option mentioned in the Wayland docs themselves would be to pass the game buffer out-of-band like with the MacOS embedder. I personally think that in our particular case, given Wayland's architecture, the approach I'm implementing is the right choice. [↩](https://github.com/godotengine/godot/pull/107435#user-content-fnref-canonical-approach-99f1fb4e3030f8288e8c8dcf46e10ade)

Fireye04 commented:
@Riteo Can confirm fix is working on my end! Thanks!

JamesMowery commented:
Any hope of this making it to the 4.5 release? :pray:

I know there was mention of a few exceptions of features that could be added last minute on a prior blog post regarding 4.5. Really hoping this is one of those exceptions! I recently switched to Wayland and not having the game embedding after having experienced it is so painful.

Appreciate the amazing work as always!

ShalokShalom commented:
Interesting! I know its done now, and it make no sense anymore to consider it now, but did you initially consider [aquamarine](https://github.com/hyprwm/aquamarine?tab=readme-ov-file#aquamarine) to replace wlroots with? 

Its used in the Hyprland WM.

Yagich commented:
tested briefly on KWin 6.3.6 with a fresh project, works great.

running a project for the first time printed an error in the debugger:
```
ERROR: Formatting error in string "Setting libdecor frame visibility to %d": a number is required.
   at: vformat (./core/variant/variant.h:922)
```
(log for this one is available if you need it)

in the 3D input override mode, first person (holding rmb) controls like WASD do not seem to work, but the camera does rotate with the mouse. (update: trying it again, it works now...)

object picking works in both 2D and 3D input override modes. always-on-top is respected, and the issue of the "internal" window being resized when moved to a different window or out of "visible" bounds that's present on the x11 driver does not exist.

when the running game is closed, an error is printed to the output: `ERROR: platform/linuxbsd/wayland/embedder.cpp:712 - Condition "p_id == INVALID_ID" is true. Returning: ERR_INVALID_PARAMETER`

however, i can't seem to edit the project normally. most times trying to edit it from the project manager crashes, and on restart it prompts to run it in recovery mode, which works every time. reloading the project from recovery mode has a 50/50 chance of working. i'm not sure if this is an issue in this PR or something that appears on master atm.

Yagich commented:
> Maybe you just forgot to focus the embedded game? You need to click it once with the left mouse button first, or the embedded game won't have keyboard focus. You can tell that it's focused if the outline of the game workspace is blue.

indeed, that was it!

here are some verbose logs: 
[editing.tar.gz](https://github.com/user-attachments/files/22435617/editing.tar.gz)
`editing1.log` is me failing to open the project.
`editing2.log` is opening the project in recovery mode, restarting successfully, loading a 3d scene and doing a few camera override things, then closing the running game with the WM's close button, and finally closing the editor the same way. in this one, the `send_wayland_message` error appears, as well as the formatting error, and another error that i *thought* i saw but couldn't reliably reproduce: `ERROR: Can't read message header: Connection reset by peer at: handle_sock (platform/linuxbsd/wayland/embedder.cpp:2531)`. hope these help!

sorry to say i can't help with the tablet support testing as i don't have one!

and lastly, thank you for the huge amount of work you've done! :)

Yagich commented:
sorry for taking a while to get back to you!

i got the latest artifact, and the invalid id issue is indeed fixed.

there is a new problem i encountered, but wasn't able to reproduce. i was in the floating workspace, and the "fake window" wasn't on top of the "parent window", and would act like a normal window regarding focus. unfortunately i didn't have the logs going when i launched it, and after a restart to try and repro it it fixed itself seemingly permanently :weary:

deralmas commented:
> Any chance of this being moved to a "ready to be merged" type state so we can test this out in the official dev builds?

Well, it's not ready to be merged yet, so not really :P

I think it's reasonably stable now, but there was the weird tablet thing and I haven't found a chance yet to look into it unfortunately.

> Someone on reddit mentioned that the Godot team won't consider merging this because it is in "draft" state

Indeed, it physically can't be merged from the github UI, as it's not marked as ready and it would only break things to do it now.

> I'm running into an issue where I can't I can't get export templates working with the version you helped me patch to get the embedding working. I'm assuming the export templates or Godot expects it to match up with whatever official version of Godot is out there.

(BTW you can now switch to the head of this branch again, I included that patch here since you reported no problems)

Well, since you're running dev builds it's probably looking for that on the server. There are various things you could do, such as [building and using your own templates](https://docs.godotengine.org/en/stable/engine_details/development/compiling/compiling_for_linuxbsd.html#building-export-templates), you're working on `4.6.dev` now, after all. The other option would be to extracting [the closest official dev templates](https://downloads.godotengine.org/?version=4.6&flavor=dev1&slug=export_templates.tpz&platform=templates (rn [dev1](https://godotengine.org/download/archive/4.6-dev1/), gotta rebase) and specifying them in the export menu.

> But beyond that, it would be great to have this in the latest dev build so I can try breaking all of Godot and reporting bugs as well.

I'd love that too but there are still some missing things. We also depend on #111493, which hasn't been merged upstream yet.

In the meantime, if you find any bug that looks unrelated to this PR, you can always double-check on a "vanilla" branch (always backup!) and report as usual. If the bug is proxy-related (or looks like it), feel free to report in this PR :D

You'll have to be patient. I'm a student working on this in my free time whenever I feel like it. I also have some long-standing personal issues slowing me down significantly.

> Either way, much appreciated as always!

Thank you!

JamesMowery commented:
@deralmas Ah that figures. The person on reddit who commented (also had a contributor badge) sort of made it sound like this was ready to and just needed the status change to be merged. I should've known better, lol!

I'll definitely check out building the templates manually, as it doesn't seem too difficult. That should hold me over. I suppose I can just do a git pull and get all the latest stuff as well, so I'll also give that a shot! I'll report back if I find any other bugs/issues and then can keep up with the newest things.

Hope the studies go well and everything else goes better. You're the Godot savior for all things Linux!

deralmas commented:
@JamesMowery I should've fixed the focus issue. I have tested it very little so please tell me if that works for you.

deralmas commented:
After some further testing, I think I'm going to leave tablet support for another time.

It looks like our tablet implementation made some weird assumptions, and would require some tender caring love in the first place. This, compounded with the somewhat quirky nature of some compositors makes it a small project of its own, which would not make sense to follow right now.

This means that, for now, running the editor (and the editor only) with the embedder enabled will disable tablet input support.

To be clear: anything that isn't the editor itself, including the project manager, exported games and even unembedded games ran from the editor, will work exactly as they always did.

In the meantime, if you really need tablet support but can go without embedding, I'll add an environment variable to disable the embedding layer, restoring things as they always were.

deralmas committed:
Wayland: Implement game embedding

This patch introduces a new protocol proxy, which multiplxes Wayland
clients into a single connection, allowing us to redirect calls (e.g.
create toplevel -> create subsurface). Mixed with some state tracking
and emulation, we can embed a full-featured client into the editor.

YeldhamDev commented:
Functionality-wise seems to work fine, but I get these errors:
```
W 0:00:00:058   init: Can't obtain the XDG decoration manager. Libdecor will be used for drawing CSDs, if available.
  <C++ Source>  platform/linuxbsd/wayland/wayland_thread.cpp:4622 @ init()
W 0:00:00:058   init: FIFO protocol not found! Frame pacing will be degraded.
  <C++ Source>  platform/linuxbsd/wayland/wayland_thread.cpp:4639 @ init()
W 0:00:00:058   init: xdg-toplevel-icon protocol not found! Cannot set window icon.
  <C++ Source>  platform/linuxbsd/wayland/wayland_thread.cpp:4643 @ init()
```
The taskbar icons also fallback to the generic Wayland logo.

YeldhamDev commented:
Testing more, sometimes it does not work correctly, it creates a separate window that is not embedded:
<img width="1242" height="710" alt="image" src="https://github.com/user-attachments/assets/d96817bd-36f2-4309-91cf-79da83549afa" />

There are also times that closing the game crashes the editor.

deralmas commented:
Thank you for testing!

> Functionality-wise seems to work fine, but I get these errors:

Do you get them when you open the editor or when you run the game? The first two protocols (FIFO and decoration) should be passed to the editor, although after rebasing I forgot to register the icon protocol for the editor proxy.

> Testing more, sometimes it does not work correctly, it creates a separate window that is not embedded:

I see an X.Org icon. Have you enabled "Prefer Wayland" in the editor settings?

> There are also times that closing the game crashes the editor.

Could you please send me the error? I'm afraid I might have seen it but it seemed pretty random. If it's a "object not found" error, that might require recompiling with the debug macro defined to figure out what's going on.

deralmas commented:
> The errors happen when running the project

Mhhh, that's to be kinda expected. Those are actually warnings, telling the user that certain features are missing because the compositor does not expose them. FIFO and decoration manager should already be there on the latest versions of KDE, but the decorations and icon protocol will always be not exposed to embedded clients.

I've also pushed a patch that exposes the ~~decoration~~ icon protocol to the editor, although it unfortunately won't solve the warnings. **Edit**: I always mix those two up, don't ask me why.

> Yes. When that bug happens, the taskbar icons have the proper icons, maybe related?

The weird thing is that it should default to a Wayland icon, not a X.Org one. Additionally, the bug you're talking about is what happens when either end is an X11 window, mostly the window to be embedded. A few commits ago I introduced a mechanism that selects the Wayland driver for embedded windows. Maybe the game is failing to connect to the socket and falls back to the X11 driver, causing this issue? I'm really not sure.

To add to the confusion, the window with the funky icons is the editor's one. The latest commit should allow it to set its own icon again but if the editor is running under Wayland, then why is it sporting an X11 icon? Maybe looking at the logs could help, `--verbose` should talk about it.

>  On those moments I weep because the artifact binaries don't come with debug symbols. 🫠 Just give me a moment and I will compile it myself.

Thank you for your patience! Building also gives you access to the `WAYLAND_EMBED_DEBUG_LOGS_ENABLED` macro, which you can set and get the densest log you've ever seen. If the error refers to an "object", it will be the only way to figure out what's going on.

> Got the backtrace:

Thank you! It looks like, probably during deinitialization, the parent window handle is invalid (or so I hope, I see nothing invalidating their reference but they also don't really have a reason to corrupt). I can't really replicate this easily, so I hope that the changes that I just pushed fix the issue.

deralmas commented:
> Yes. When that bug happens, the taskbar icons have the proper icons, maybe related?

Oh. I've just re-read this. For some reason the editor was running under X11 backend then, as the icon protocol was not exposed when you tested this and it showed an X11 icon. Did the logs complain about something? It looks like the editor fell back to the X11 backend for whatever reason.

deralmas commented:
Additionally I've pushed a change that runs all embedded clients with the X11 backend when running under X11. Hopefully this means that if there's an issue at least embedding won't outright break. Note that if for some reason only the embedded game fails and falls back into the alternative driver, embedding will look wonky again but improving this would require deeper changes into how and whether falling back happens.

YeldhamDev commented:
Another crash:
```
ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
   at: propagate_notification (scene/main/node.cpp:2597)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.6.dev.custom_build (b5c01e7119fa5153e794465b8f3afdb1c76f426f)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x1a070) [0x7fb058f45070] (??:0)
[2] WaylandEmbedder::LocalObjectHandle::is_valid() const (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.h:226)
[3] WaylandEmbedder::handle_request(WaylandEmbedder::LocalObjectHandle, unsigned int, unsigned int const*, unsigned long) (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:1624 (discriminator 1))
[4] WaylandEmbedder::handle_msg_info(WaylandEmbedder::Client*, WaylandEmbedder::msg_info const*, unsigned int*, int*) (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:2408 (discriminator 1))
[5] WaylandEmbedder::handle_sock(int) (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:2718 (discriminator 2))
[6] WaylandEmbedder::handle_fd(int, int) (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:2900)
[7] WaylandEmbedder::poll_sockets() (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:680 (discriminator 4))
[8] WaylandEmbedder::_thread_loop(void*) (/var/home/michael/Downloads/godot-wl-proxy/platform/linuxbsd/wayland/wayland_embedder.cpp:2735)
[9] Thread::callback(unsigned long, Thread::Settings const&, void (*)(void*), void*) (/var/home/michael/Downloads/godot-wl-proxy/core/os/thread.cpp:66)
[10] void std::__invoke_impl<void, void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>(std::__invoke_other, void (*&&)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/include/c++/15/bits/invoke.h:63)
[11] std::__invoke_result<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>::type std::__invoke<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>(void (*&&)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/include/c++/15/bits/invoke.h:99)
[12] void std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (/usr/include/c++/15/bits/std_thread.h:303)
[13] std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::operator()() (/usr/include/c++/15/bits/std_thread.h:310)
[14] std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> > >::_M_run() (/usr/include/c++/15/bits/std_thread.h:255)
[15] /var/home/michael/Downloads/godot-wl-proxy/bin/godot.linuxbsd.editor.dev.x86_64() [0x6666b24] (thread.o:?)
[16] /lib64/libc.so.6(+0x71f54) [0x7fb058f9cf54] (??:0)
[17] /lib64/libc.so.6(+0xf532c) [0x7fb05902032c] (??:0)
-- END OF C++ BACKTRACE --
================================================================
```

deralmas commented:
@YeldhamDev I've hopefully fixed the issue. I tried replicating it on KWin before touching the code further but I could not. The annoying thing is that I'm not exactly sure what's going on, so I can't be sure that I fixed the root problem for good. Please tell me if you still get those crashes when you close the editor.

deralmas commented:
Thank you for testing. This is pretty much the same issue xD

I'll fix it ASAP.

YeldhamDev commented:
Huh, I managed to get the following crash **_once_** and then not anymore:
```
wl_display#1: error 0: invalid object 168
ERROR: Wayland client error code 22.
   at: _poll_events_thread (platform/linuxbsd/wayland/wayland_thread.cpp:3176)
ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
   at: propagate_notification (scene/main/node.cpp:2597)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.6.dev.custom_build (6a1917cd8473bf4c4a7a936811c6885fe6acefb6)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
ERROR: Wayland client error code 22.
   at: wait_frame_suspend_ms (platform/linuxbsd/wayland/wayland_thread.cpp:5101)
```
Maybe apply some counter-measure for this specific code path if possible, and then I think we can consider the code crash-free.

deralmas commented:
@YeldhamDev all right. The log is not exactly the same error as last time but I wouldn't surprised if things were related. Apparently, restarting the client a lot of times will not clean up some objects. When they trigger an event, we ignore those "phantom objects" and everything falls apart.

I've added some fallback logic to keep the show running, with some one-time warnings to let you know when the thing happens.

The fix is good behavior, it's perfectly stable and I should've added in the first place, but every "phantom object" is a small leak and should not happen. I'll keep debugging.

Thank you so much for your patience testing this thing and please let me know if you find any other crash or weird behavior.

deralmas commented:
Oh, then something is wrong. Looking at the logs, it looks like I missed the manager from the list :sweat_smile:

Hopefully this new commit fixes the issue.

YeldhamDev reviewed (approved):
Functionality-wise everything seems to work. But still needs approval from someone who better understands Wayland code.

AThousandShips commented on platform/linuxbsd/wayland/wayland_embedder.cpp:
```diff
@@ -0,0 +1,2973 @@
+	const uint32_t *body = msg_data + 2;
+	//size_t body_len = msg_len - (WL_WORD_SIZE * 2);
+
+	// FIXME: Make sure that it makes sense to track this protocol. Not only it is
```
```suggestion
	// FIXME: Make sure that it makes sense to track this protocol. Not only is it
```

AThousandShips commented on platform/linuxbsd/wayland/wayland_embedder.h:
```diff
@@ -0,0 +1,577 @@
+#ifdef TOOLS_ENABLED
+
+#include "core/error/error_list.h"
+#include "core/error/error_macros.h"
```
```suggestion
```
Already included

AThousandShips commented on platform/linuxbsd/wayland/wayland_embedder.h:
```diff
@@ -0,0 +1,637 @@
+#define EMBED_ANCILLARY_BUF_SIZE 4096
+
+class WaylandEmbedder {
+private:
```
```suggestion
```

AThousandShips commented on platform/linuxbsd/wayland/wayland_embedder.h:
```diff
@@ -0,0 +1,637 @@
+	static Error send_raw_message(int p_socket, std::initializer_list<struct iovec> p_vecs, const LocalVector<int> &p_fds = LocalVector<int>());
+
+	static Error send_wayland_message(int p_socket, uint32_t p_id, uint32_t p_opcode, const uint32_t *p_args, const size_t p_args_words);
+	static Error send_wayland_message(ProxyDirection p_direction, int p_socket, uint32_t p_id, const struct wl_interface &p_interface, uint32_t p_opcode, LocalVector<union wl_argument> p_args);
```
```suggestion
	static Error send_wayland_message(ProxyDirection p_direction, int p_socket, uint32_t p_id, const struct wl_interface &p_interface, uint32_t p_opcode, const LocalVector<union wl_argument> &p_args);
```

AThousandShips commented on platform/linuxbsd/wayland/wayland_embedder.h:
```diff
@@ -0,0 +1,637 @@
+		return send_wayland_message(ProxyDirection::COMPOSITOR, p_socket, p_id, p_interface, p_opcode, p_args);
+	}
+
+	static Error send_wayland_event(int p_socket, uint32_t p_id, const struct wl_interface &p_interface, uint32_t p_opcode, LocalVector<union wl_argument> p_args) {
```
```suggestion
	static Error send_wayland_event(int p_socket, uint32_t p_id, const struct wl_interface &p_interface, uint32_t p_opcode, const LocalVector<union wl_argument> &p_args) {
```

AThousandShips commented on platform/linuxbsd/wayland/display_server_wayland.cpp:
```diff
@@ -1505,6 +1508,94 @@ bool DisplayServerWayland::get_swap_cancel_ok() {
+			toplevel = libdecor_frame_get_xdg_toplevel(ws->libdecor_frame);
+		}
+
+		ERR_FAIL_NULL_V(ws, ERR_BUG);
```
Needs to be moved up as it's used on line 1537

AThousandShips commented:
I think they can be separated into a dedicated 4.5 branch and then that can be cherry picked, I think this PR makes sense as one unit

akien-mga reviewed (commented):
Tested briefly on KDE Wayland (6.5.2), seems to work great!

I haven't reviewed the Wayland code in depth, just skimmed through the rest of the changes.

bruvzg reviewed (approved):
Seems to be working as expected.

bruvzg commented:
Got editor to crash when stopping running project, but not reliably reproducible:
```
ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
   at: propagate_notification (scene/main/node.cpp:2597)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.6.dev.custom_build (02c2f268d1962b9fbd9b5cb0b7e3fa7ff9e3d526)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x1a070) [0x7fa5ebf33070] (??:0)
[2] ./bin/godot.linuxbsd.editor.x86_64() [0x546a78] (./platform/linuxbsd/wayland/wayland_embedder.cpp:1647 (discriminator 1))
[3] ./bin/godot.linuxbsd.editor.x86_64() [0x54974b] (./platform/linuxbsd/wayland/wayland_embedder.cpp:2436 (discriminator 1))
[4] ./bin/godot.linuxbsd.editor.x86_64() [0x54a35a] (./platform/linuxbsd/wayland/wayland_embedder.cpp:2750)
[5] ./bin/godot.linuxbsd.editor.x86_64() [0x54ad30] (./platform/linuxbsd/wayland/wayland_embedder.cpp:2932)
[6] ./bin/godot.linuxbsd.editor.x86_64() [0x54aeda] (./core/templates/local_vector.h:76)
[7] ./bin/godot.linuxbsd.editor.x86_64() [0x54b0c8] (/usr/include/c++/15/bits/atomic_base.h:501)
[8] ./bin/godot.linuxbsd.editor.x86_64() [0x572cec5] (./core/os/thread.cpp:64)
[9] ./bin/godot.linuxbsd.editor.x86_64() [0x6213aa4] (thread.o:?)
[10] /lib64/libc.so.6(+0x71f54) [0x7fa5ebf8af54] (??:0)
[11] /lib64/libc.so.6(+0xf532c) [0x7fa5ec00e32c] (??:0)
-- END OF C++ BACKTRACE --
================================================================
```

deralmas commented on platform/linuxbsd/wayland/wayland_thread.cpp:
```diff
@@ -4445,7 +4570,52 @@ Error WaylandThread::init() {
+
+		// Debug
+		if (OS::get_singleton()->get_environment("GODOT_DEBUG_EMBEDDER_SINGLE_INSTANCE") == "1") {
+			print_line("Pausing as per GODOT_DEBUG_EMBEDDER_SINGLE_INSTANCE.");
```
Yeah fair, I don't really see an user running it with the environment flag as a normal diagnostic tool. I'll remove it

Calinou reviewed (commented):
Tested locally, it works as expected.

```
Operating System: Fedora Linux 43
KDE Plasma Version: 6.5.2
KDE Frameworks Version: 6.19.0
Qt Version: 6.10.0
Kernel Version: 6.17.7-300.fc43.x86_64 (64-bit)
Graphics Platform: Wayland

Processors: 32 × AMD Ryzen 9 9950X3D 16-Core Processor
Memory: 64 GiB of RAM (62.3 GiB usable)
Graphics Processor: NVIDIA GeForce RTX 5090
Manufacturer: Micro-Star International Co., Ltd.
Product Name: MS-7E49
System Version: 1.0
```

### Prefer Wayland disabled (= XWayland)

*Notice the delay in resizing. Also, when alt-tabbing, the game window overlaps the newly focused application until you release the <kbd>Alt</kbd> key.*

[Screencast_20251119_195058.webm](https://github.com/user-attachments/assets/0733a834-105c-42bb-9e5c-45a857855acd)

### Prefer Wayland enabled (= this PR)

*Neither of the issues mentioned above occur.*

[Screencast_20251119_195220.webm](https://github.com/user-attachments/assets/9c00efd0-3449-4471-81e9-db1f6d94b5df)

One issue I noticed (and that didn't occur in XWayland) is that when I choose the 3D mode and override the game camera, I can't move the camera until I left-click inside the game window to ensure the game window gains focus (as indicated by the blue outline). (Right-clicking has no effect on focus, even if the blue outline is already visible, although it does allow freelook rotation to work.) This also occurs if I click/alt-tab away from the editor and back into it.

In XWayland, the window gains focus as soon as the mouse moves inside the window's bounds instead.

I can't reproduce the sporadic crash when stopping the project, either by pressing <kbd>F8</kbd> or using the stop button in the editor (I've tried this around 20 times).

akien-mga reviewed (approved):
Let's go!

JamesMowery commented:
Yay yay yay! Super excited to try the new changes! Thanks so much for all the amazing work!

Fireye04 commented:
Holy shit we're merged! Congrats and thanks a bunch @deralmas!

ArchercatNEO commented on platform/linuxbsd/wayland/display_server_wayland.cpp:
```diff
@@ -1505,6 +1508,94 @@ bool DisplayServerWayland::get_swap_cancel_ok() {
+			toplevel = libdecor_frame_get_xdg_toplevel(ws->libdecor_frame);
+		}
+
+		ERR_FAIL_NULL_V(toplevel, ERR_CANT_CREATE);
```
We're using `libdecor` without an `#ifdef LIBDECOR_ENABLED` flag here.

ndbn commented:
`scons target=editor production=no compiledb=yes debug_symbols=yes dev_mode=yes`

<img width="1143" height="175" alt="image" src="https://github.com/user-attachments/assets/def24a98-28ef-4dd8-a86d-5b9665cc9429" />

deralmas commented:
Hi @ndbn thank you for reporting this compile errror. Dunno why no-one caught this warning before, warnings in general seem somewhat flaky, in my experience :sweat_smile:

It makes sense, as both branches only contain the debug macro, which gets stubbed out with an `#ifdef`, so they're identical indeed.

I'll make a PR soon.

deralmas commented:
Oh, sorry about that. I completely missed it, or maybe just forgot about it :sweat_smile:

eobet commented:
Sorry if this is the wrong place for this comment, but from a user perspective I find it weird that this requires "single window mode" to be off when it appears to not use a second window (unlike what I've been forced to use on Linux for the last year where the game opens in a new window).

I find this problematic not just because it's illogical (from a UX perspective) but because window sizes seem to be all over the place in Godot with Wayland and display scaling (whereas in single window mode, this was not an issue).

deralmas commented:
Hi @eobet. Nothing stops us from embedding in single-window mode, technically. Actually, nothing technically stops us from embedding in single-window mode on X11/Windows too.

Originally (before Wayland embedding), it was enabled in single-window mode too, but it got disabled in #101936.

The issue is, in both cases, that the embedded game covers popups and the like. While we don't use a different window, we use a different window "layer" (`wl_subsurface`), which still composites over any eventual popups.

We use a subsurface to avoid reading and re-compositing everything, which would be a _lot_ more complex, as mentioned in this PR description.

> but because window sizes seem to be all over the place in Godot with Wayland and display scaling (whereas in single window mode, this was not an issue).

This is a bug, and something that _will_ be fixed. There are various interconnected things at play, but I've got a workaround ready, and improvements to the engine API are in the works. You can track #110643 and perhaps give its linked PR a shot.

Please note that the Wayland backend is marked as experimental and is still not polished as we'd like. That's why it's currently hidden behind an opt-in switch.

Feel free to ask if you've got further questions, I'm happy to help.

mahkoh commented:
@deralmas: FYI I've taken your idea and turned it into a rust crate: https://lore.freedesktop.org/wayland-devel/CAHijbEUuMG9-dCPFrDRMrEzB60YgCbjdeRDgUZaZV3JyMfJ-0w@mail.gmail.com/T/#u

Just wanted to let you know that I thought this was pretty neat.

musjj commented:
@JamesMowery

Hey did you ever get this feature to work on Hyprland, Niri or other tiling compositors? I already set both my editor and game to Wayland, but I still can't get this embedded window feature to work.

JamesMowery commented:
@musjj Working quite fine for me on v4.6.1-rc1. This is on Niri. Be sure to uncheck the option I have highlighted.

<img width="1814" height="1104" alt="image" src="https://github.com/user-attachments/assets/71731fac-39cb-40da-b454-0796989ca5f3" />

orowith2os commented:
I'm curious if @deralmas would be interested in splitting off the Wayland interception code into a separate library? I'd like to use the work to do almost this exact thing for my own use cases, and having a C/C++ implementation (Rust is too annoying for me to use with this) would help a lot.

deralmas commented:
Hi @orowith2os, yes, I'm definitely interested in turning this into a library!

Please note that currently it's pretty hacky (see the PR description for more context), uses a lot of internal (but public) libwayland types (`wl_argument`, `wl_interface`...), and depends on Godot's custom STD library. I also have to juggle other things too, like fixing various Godot Wayland bugs, so it's going to take a while to do properly.

If you're working on something right now, it might be easier to bundle the required Godot templates and use `godot_embedder.{cpp,h}` directly. Since it's a thread and can be interacted with through Wayland requests I think that you can make it work as-is with minimal modifications. Not ideal, but hopefully does the job for the meantime.

Please feel free to contact me directly if you need anything :D

deralmas commented:
Yes you did! Please redact that though, due to spam-scraping concerns :P
